File: backend/src/automation/linkedin.py
# backend/src/automation/linkedin.py

from .base_automation import BaseAutomation

class LinkedInAutomation(BaseAutomation):
    async def setup(self):
        await super().setup()
        await self.page.goto("https://www.linkedin.com/")

    async def login(self, username, password):
        await self.page.goto("https://www.linkedin.com/login")
        await self.random_sleep(2, 3)
        await self.page.fill('#username', username)
        await self.page.fill('#password', password)
        await self.page.click('button[type="submit"]')
        await self.page.wait_for_navigation()

    async def send_dm(self, username, message):
        try:
            await self.page.goto("https://www.linkedin.com/messaging/")
            await self.random_sleep(2, 3)
            
            await self.page.click('button[aria-label="Compose message"]')
            await self.random_sleep(1, 2)
            
            await self.page.fill('input[aria-label="Type to search for connections"]', username)
            await self.random_sleep(1, 2)
            
            await self.page.click(f'div[aria-label="Search results"] li:has-text("{username}")')
            await self.random_sleep(1, 2)
            
            personalized_message = self.personalize_message(message, username)
            await self.page.fill('div[role="textbox"]', personalized_message)
            await self.random_sleep(1, 2)
            
            await self.page.click('button[aria-label="Send"]')
            await self.random_sleep(2, 3)
            
            return True
        except Exception as e:
            print(f"Error sending message to {username} on LinkedIn: {str(e)}")
            return False

    async def logout(self):
        await self.page.click('button[data-control-name="nav.settings"]')
        await self.random_sleep(1, 2)
        await self.page.click('a[data-control-name="nav.settings_signout"]')

File: dms/lib/linkedin.ts
// lib/linkedinAutomation.ts
import { chromium, Browser, Page } from 'playwright';

export async function sendLinkedInDM(email: string, password: string, recipient: string, message: string) {
  const browser: Browser = await chromium.launch({ headless: false });
  const page: Page = await browser.newPage();

  try {
    // Login to LinkedIn
    await page.goto('https://www.linkedin.com/login');
    await page.fill('input#username', email);
    await page.fill('input#password', password);
    await page.click('button[type="submit"]');

    // Wait for login to complete
    await page.waitForNavigation();

    // Navigate to messaging
    await page.goto('https://www.linkedin.com/messaging/');

    // Click on "New message" button
    await page.click('button[aria-label="Compose message"]');

    // Search for recipient
    await page.fill('input[aria-label="Type a name or multiple names"]', recipient);
    await page.click(`div[aria-label="${recipient}"]`);

    // Send message
    await page.fill('div[role="textbox"]', message);
    await page.click('button[aria-label="Send"]');

    console.log(`Message sent to ${recipient} on LinkedIn successfully.`);
  } catch (error) {
    console.error('Error occurred:', error);
  } finally {
    await browser.close();
  }
}